akka-stream-tests/src/test/scala/akka/stream/impl/ActorInterpreterSpec.scala

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl
+
+import akka.stream.Supervision._
+import akka.stream.testkit.AkkaSpec
+import akka.stream._
+import akka.stream.scaladsl._
+import akka.stream.testkit.StreamTestKit._
+import akka.stream.impl.fusing.ActorInterpreter
+import akka.stream.stage.Stage
+import akka.stream.stage.PushPullStage
+import akka.stream.stage.Context
+import akka.testkit.TestLatch
+import scala.concurrent.Await
+import scala.concurrent.duration._
+
+class ActorInterpreterSpec extends AkkaSpec {
+  import FlowGraph.Implicits._
+
+  implicit val mat = ActorFlowMaterializer()
+
+  class Setup(ops: List[Stage[_, _]] = List(fusing.Map({ x: Any ⇒ x }, stoppingDecider))) {
+    val up = PublisherProbe[Int]
+    val down = SubscriberProbe[Int]
+    private val props = ActorInterpreter.props(mat.settings, ops).withDispatcher("akka.test.stream-dispatcher")
+    val actor = system.actorOf(props)
+    val processor = ActorProcessorFactory[Int, Int](actor)
+  }
+
+  "An ActorInterpreter" must {
+
+    "pass along early cancellation" in new Setup {
+      processor.subscribe(down)
+      val sub = down.expectSubscription()
+      sub.cancel()
+      up.subscribe(processor)
+      val upsub = up.expectSubscription()
+      upsub.expectCancellation()
```

Comment:
This currently fails, because I fixed ActorInterpreter to actually shut down upon cancellation, and then there is nobody to generate the upstream’s `cancel()` call. This is the reason for #16986.